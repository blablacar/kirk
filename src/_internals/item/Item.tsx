import React from 'react'
import cc from 'classcat'

import { color } from '../../_utils/branding'
import { A11yProps, pickA11yProps } from '../../_utils/interfaces'
import { Button } from '../../button/Button'
import { ChevronIcon } from '../../icon/chevronIcon'
import { NormalizeProps } from '../../layout/layoutNormalizer'
import { Text, TextDisplayType, TextTagType } from '../../text'

export enum ItemStatus {
  DEFAULT = 'default',
  LOADING = 'loading',
  CHECKED = 'checked',
}

export type ItemProps = A11yProps &
  NormalizeProps &
  Readonly<{
    chevron?: boolean
    className?: string
    href?: string | JSX.Element
    highlighted?: boolean
    isClickable?: boolean
    isWrappable?: boolean
    leftTitle?: React.ReactNode
    leftTitleButtonAddon?: React.ReactElement<Button>
    leftTitleDisplay?: TextDisplayType
    leftTitleColor?: string
    leftBody?: string | React.ReactNode
    leftBodyDisplay?: TextDisplayType
    leftBodyColor?: string
    leftBodyAnnotation?: React.ReactNode
    leftBodyAnnotationDisplay?: TextDisplayType
    leftBodyAnnotationColor?: string
    leftAddon?: React.ReactNode
    rightTitle?: string | JSX.Element
    rightTitleDisplay?: TextDisplayType
    rightTitleStrikeThrough?: boolean
    rightTitleAriaProps?: A11yProps
    rightTitleColor?: string
    rightBody?: React.ReactNode
    rightBodyDisplay?: TextDisplayType
    rightBodyColor?: string
    rightAddon?: React.ReactNode
    tag?: JSX.Element
    onClick?: (event: React.MouseEvent<HTMLElement>) => void
    onBlur?: (event: React.FocusEventHandler<HTMLElement>) => void
    onFocus?: (event: React.FocusEventHandler<HTMLElement>) => void
    onMouseDown?: (event: React.MouseEvent<HTMLElement>) => void
    hideHoverBackground?: boolean
    disabled?: boolean
  }>

export const Item = (props: ItemProps) => {
  const {
    chevron,
    className,
    href,
    onClick,
    onBlur,
    onFocus,
    onMouseDown,
    highlighted,
    isClickable,
    isWrappable,
    leftTitle,
    leftTitleButtonAddon,
    rightTitleStrikeThrough,
    rightTitleAriaProps,
    leftTitleDisplay = TextDisplayType.TITLE,
    leftTitleColor,
    leftBody,
    leftBodyDisplay = TextDisplayType.BODY,
    leftBodyColor,
    leftBodyAnnotation,
    leftBodyAnnotationDisplay,
    leftBodyAnnotationColor,
    leftAddon,
    rightTitle,
    rightTitleDisplay = TextDisplayType.TITLE,
    rightTitleColor,
    rightBody,
    rightBodyDisplay = TextDisplayType.BODY,
    rightBodyColor,
    rightAddon,
    tag = <div />,
    hideHoverBackground = false,
    disabled = false,
  } = props
  const a11yAttrs = pickA11yProps<ItemProps>(props)
  let Tag = tag.type
  let tagProps = tag.props

  if (href) {
    if (typeof href !== 'string') {
      Tag = href.type
      tagProps = { ...tagProps, ...href.props }
    } else {
      Tag = 'a'
      tagProps = { href }
    }
  }

  if (Tag === 'button') {
    // The code below make sure that if we use a HTML button, a proper type='button' or
    // type='submit' is added. type='button' is the default.
    // This type=button is needed to make sure non-submit buttons are not activated when
    // submitting forms.
    // See: https://stackoverflow.com/questions/4763638/enter-triggers-button-click
    const buttonType = tag.props.type || 'button'
    tagProps = { ...tagProps, type: buttonType }
  }

  const hasRightText = rightTitle || rightBody

  const getTextColor = (textColor: string) => (disabled ? color.gray : textColor)

  const getLeftTitle = (value: React.ReactNode): React.ReactNode => {
    if (React.isValidElement(value)) {
      return leftTitle
    }
    return (
      <Text
        className={leftTitleButtonAddon ? 'kirk-item-title--withButtonAddon' : null}
        display={leftTitleDisplay}
        textColor={getTextColor(leftTitleColor)}
        tag={TextTagType.SPAN}
      >
        {leftTitle}
        {/* In case of a clickable Item don't display the addon button */}
        {!href && leftTitleButtonAddon}
      </Text>
    )
  }

  return (
    <Tag
      {...tagProps}
      onClick={onClick}
      onFocus={onFocus}
      onBlur={onBlur}
      onMouseDown={onMouseDown}
      className={cc([
        'kirk-item',
        {
          'kirk-item--highlighted': highlighted,
          'kirk-item--clickable': isClickable,
          'kirk-item--wrappable': isWrappable,
          'kirk-item--hideHoverBackground': hideHoverBackground,
        },
        className,
      ])}
      {...a11yAttrs}
    >
      <span className="kirk-item-leftWrapper">
        {leftAddon && <span className="kirk-item-leftAddon">{leftAddon}</span>}
        <span className="kirk-item-leftText">
          {leftTitle && getLeftTitle(leftTitle)}
          {leftBody && (
            <Text
              className="kirk-item-body"
              display={leftBodyDisplay}
              textColor={getTextColor(leftBodyColor)}
              tag={TextTagType.SPAN}
            >
              {leftBody}
            </Text>
          )}
          {leftBodyAnnotation && (
            <Text
              className="kirk-item-body-annotation"
              display={leftBodyAnnotationDisplay}
              textColor={getTextColor(leftBodyAnnotationColor)}
              tag={TextTagType.SPAN}
            >
              {leftBodyAnnotation}
            </Text>
          )}
        </span>
      </span>
      {hasRightText && (
        <span className="kirk-item-rightText">
          {rightTitle && (
            <Text
              className={cc([
                'kirk-item-title',
                {
                  'kirk-item--strikethrough': rightTitleStrikeThrough,
                },
              ])}
              display={rightTitleDisplay}
              textColor={getTextColor(rightTitleColor)}
              tag={TextTagType.DIV}
              {...rightTitleAriaProps}
            >
              {rightTitle}
            </Text>
          )}
          {rightBody && (
            <Text
              className="kirk-item-body"
              display={rightBodyDisplay}
              textColor={getTextColor(rightBodyColor)}
              tag={TextTagType.DIV}
            >
              {rightBody}
            </Text>
          )}
        </span>
      )}
      {rightAddon && <span className="kirk-item-rightAddon">{rightAddon}</span>}
      {chevron && (
        <span className="kirk-item-rightAddon">
          <ChevronIcon iconColor={!isClickable ? color.gray : color.lightMidnightGreen} />
        </span>
      )}
    </Tag>
  )
}
