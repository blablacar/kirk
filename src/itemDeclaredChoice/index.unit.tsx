import React from 'react'
import { shallow, mount } from 'enzyme'
import renderer from 'react-test-renderer'

import Item from '_utils/item'
import ItemDeclaredChoice, { ItemDeclaredChoiceProps } from './index'

describe('ItemDeclaredChoice', () => {
  const defaultProps: ItemDeclaredChoiceProps = {
    label: 'Label',
    name: 'name',
    value: 0,
    className: 'custom-class-name',
    labelTitle: 'Label title',
    data: 'Data',
    dataInfo: 'Data info',
    checked: false,
    disabled: false,
    onChange() {},
    status: ItemDeclaredChoice.STATUS.DEFAULT,
    key: 0,
  }

  it('Should use the Item component', () => {
    const itemDeclaredChoice = shallow(<ItemDeclaredChoice {...defaultProps} />)
    expect(itemDeclaredChoice.find(Item).exists()).toBe(true)
  })
  it('Should forward its props to the Item component', () => {
    const itemDeclaredChoice = renderer.create(<ItemDeclaredChoice {...defaultProps} />).toJSON()
    expect(itemDeclaredChoice).toMatchSnapshot()
  })
  it('Should display a CheckIcon when the input is checked', () => {
    const itemDeclaredChoice = renderer
      .create(<ItemDeclaredChoice {...defaultProps} checked />)
      .toJSON()
    expect(itemDeclaredChoice).toMatchSnapshot()
  })
  it('Should display a Loader when the component is in loading status', () => {
    const itemDeclaredChoice = renderer
      .create(<ItemDeclaredChoice {...defaultProps} status={ItemDeclaredChoice.STATUS.LOADING} />)
      .toJSON()
    expect(itemDeclaredChoice).toMatchSnapshot()
  })
  it('Should bind the onChange callback of the input', () => {
    const itemDeclaredChoice = mount(<ItemDeclaredChoice {...defaultProps} />)
    expect(itemDeclaredChoice.find('input').prop('onChange')).toEqual(
      itemDeclaredChoice.instance().onChange,
    )
  })
  it('Should call the onChange prop with name and value when the input changes', () => {
    const onChangeMock = jest.fn()
    const itemDeclaredChoice = shallow(
      <ItemDeclaredChoice {...defaultProps} onChange={onChangeMock} />,
    )
    itemDeclaredChoice.instance().onChange()
    expect(onChangeMock).toHaveBeenCalledWith({
      name: defaultProps.name,
      value: defaultProps.value,
    })
  })
})
