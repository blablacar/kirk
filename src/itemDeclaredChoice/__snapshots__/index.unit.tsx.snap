// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItemDeclaredChoice Should display a CheckIcon when the input is checked 1`] = `
<label
  className="jsx-1500099655 kirk-item custom-class-name"
>
  <div
    className="jsx-1500099655 kirk-item-leftText"
  >
    <div
      className="kirk-text kirk-text-title kirk-item-title"
    >
      Label title
    </div>
    <div
      className="kirk-text kirk-text-body kirk-item-body"
    >
      Label
    </div>
  </div>
  <div
    className="jsx-1500099655 kirk-item-rightText"
  >
    <div
      className="kirk-text kirk-text-subheaderStrong kirk-item-title"
    >
      Data
    </div>
    <div
      className="kirk-text kirk-text-body kirk-item-body"
    >
      Data info
    </div>
  </div>
  <div
    className="jsx-1500099655 kirk-item-rightAddon"
  >
    <input
      checked={true}
      disabled={false}
      name="name"
      onChange={[Function]}
      type="radio"
      value={0}
    />
    <svg
      aria-hidden={true}
      className="jsx-1043659059 jsx-59167708 kirk-icon"
      height={24}
      viewBox="0 0 24 24"
      width={24}
      xmlns="http://www.w3.org/2000/svg"
    >
      
      <path
        className="jsx-1043659059 jsx-59167708"
        d="M6.5 12.5l4 4 8-8"
        fill="none"
        stroke="#FFF"
        strokeLinecap="round"
        strokeLinejoin="round"
        strokeMiterlimit="10"
        strokeWidth="1"
      />
    </svg>
  </div>
</label>
`;

exports[`ItemDeclaredChoice Should display a Loader when the component is in loading status 1`] = `
<label
  className="jsx-1500099655 kirk-item custom-class-name"
>
  <div
    className="jsx-1500099655 kirk-item-leftText"
  >
    <div
      className="kirk-text kirk-text-title kirk-item-title"
    >
      Label title
    </div>
    <div
      className="kirk-text kirk-text-body kirk-item-body"
    >
      Label
    </div>
  </div>
  <div
    className="jsx-1500099655 kirk-item-rightText"
  >
    <div
      className="kirk-text kirk-text-subheaderStrong kirk-item-title"
    >
      Data
    </div>
    <div
      className="kirk-text kirk-text-body kirk-item-body"
    >
      Data info
    </div>
  </div>
  <div
    className="jsx-1500099655 kirk-item-rightAddon"
  >
    <input
      checked={false}
      disabled={true}
      name="name"
      onChange={[Function]}
      type="radio"
      value={0}
    />
    <div
      className="jsx-2765694959 kirk-loader"
      style={
        Object {
          "height": "24px",
          "width": "24px",
        }
      }
    >
      <svg
        aria-hidden={true}
        className="jsx-2358291114 kirk-icon spinning"
        height={24}
        viewBox="0 0 66 66"
        width={24}
        xmlns="http://www.w3.org/2000/svg"
      >
        
        <circle
          className="jsx-2358291114"
          cx="33"
          cy="33"
          fill="none"
          r="30"
          stroke="#5DD167"
        />
      </svg>
    </div>
  </div>
</label>
`;

exports[`ItemDeclaredChoice Should forward its props to the Item component 1`] = `
<label
  className="jsx-1500099655 kirk-item custom-class-name"
>
  <div
    className="jsx-1500099655 kirk-item-leftText"
  >
    <div
      className="kirk-text kirk-text-title kirk-item-title"
    >
      Label title
    </div>
    <div
      className="kirk-text kirk-text-body kirk-item-body"
    >
      Label
    </div>
  </div>
  <div
    className="jsx-1500099655 kirk-item-rightText"
  >
    <div
      className="kirk-text kirk-text-subheaderStrong kirk-item-title"
    >
      Data
    </div>
    <div
      className="kirk-text kirk-text-body kirk-item-body"
    >
      Data info
    </div>
  </div>
  <div
    className="jsx-1500099655 kirk-item-rightAddon"
  >
    <input
      checked={false}
      disabled={false}
      name="name"
      onChange={[Function]}
      type="radio"
      value={0}
    />
    <svg
      aria-hidden={true}
      className="jsx-2358291114 kirk-icon thin"
      height={24}
      viewBox="0 0 66 66"
      width={24}
      xmlns="http://www.w3.org/2000/svg"
    >
      
      <circle
        className="jsx-2358291114"
        cx="33"
        cy="33"
        fill="none"
        r="30"
        stroke="#00AFF5"
      />
    </svg>
  </div>
</label>
`;
